File: ecommerce-frontend/src/services/productService.ts
import apiClient, { extractErrorMessage, ApiError } from "@/config/apiClient";

export interface Product {
  id: string;
  name: string;
  description: string;
  price: number;
  categoryId?: number;
  category?: string;
  image: string;
  stock: number;
  rating?: number;
  active?: boolean;
  createdAt?: string;
  updatedAt?: string;
}

export interface ProductFilter {
  category?: string;
  minPrice?: number;
  maxPrice?: number;
  search?: string;
  page?: number;
  limit?: number;
}

// Type guard for Product
const isValidProduct = (data: unknown): data is Product => {
  if (!data || typeof data !== "object") return false;
  const product = data as Product;
  return (
    typeof product.id !== "undefined" &&
    typeof product.name === "string" &&
    typeof product.price === "number"
  );
};

// Type guard for Product array
const isValidProductArray = (data: unknown): data is Product[] => {
  return Array.isArray(data) && (data.length === 0 || data.every(isValidProduct));
};

// Sanitize product data
const sanitizeProduct = (product: Product): Product => ({
  ...product,
  name: product.name?.trim() || "",
  description: product.description?.trim() || "",
  price: Math.max(0, product.price || 0),
  stock: Math.max(0, product.stock || 0),
  rating: product.rating ? Math.min(5, Math.max(0, product.rating)) : undefined,
  image: product.image?.trim() || "/placeholder.svg",
});

export const productService = {
  // Get all products
  getProducts: async (filters?: ProductFilter): Promise<Product[]> => {
    try {
      // Sanitize filters
      const sanitizedFilters = filters ? {
        ...filters,
        search: filters.search?.trim().substring(0, 200),
        minPrice: filters.minPrice !== undefined ? Math.max(0, filters.minPrice) : undefined,
        maxPrice: filters.maxPrice !== undefined ? Math.max(0, filters.maxPrice) : undefined,
        page: filters.page !== undefined ? Math.max(1, filters.page) : undefined,
        limit: filters.limit !== undefined ? Math.min(100, Math.max(1, filters.limit)) : undefined,
      } : undefined;

      const response = await apiClient.get<Product[]>("/products", {
        params: sanitizedFilters,
      });
      
      if (!isValidProductArray(response.data)) {
        console.warn("Invalid product data received from API");
        return [];
      }
      
      return response.data.map(sanitizeProduct);
    } catch (error) {
      console.error("Error fetching products:", extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "FETCH_PRODUCTS_ERROR"
      );
    }
  },

  // Get product by ID
  getProductById: async (id: string): Promise<Product> => {
    if (!id || id.trim() === "") {
      throw new ApiError("Product ID is required", 400, "INVALID_ID");
    }
    
    try {
      const response = await apiClient.get<Product>(`/products/${encodeURIComponent(id)}`);
      
      if (!isValidProduct(response.data)) {
        throw new ApiError("Invalid product data received", 500, "INVALID_DATA");
      }
      
      return sanitizeProduct(response.data);
    } catch (error) {
      if (error instanceof ApiError) throw error;
      console.error(`Error fetching product ${id}:`, extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "FETCH_PRODUCT_ERROR"
      );
    }
  },

  // Search products
  searchProducts: async (query: string): Promise<Product[]> => {
    try {
      // Sanitize and validate search query
      const sanitizedQuery = query?.trim().substring(0, 200) || "";
      
      if (sanitizedQuery.length === 0) {
        return productService.getProducts();
      }
      
      const response = await apiClient.get<Product[]>("/products/search", {
        params: { q: sanitizedQuery },
      });
      
      if (!isValidProductArray(response.data)) {
        console.warn("Invalid product data received from search API");
        return [];
      }
      
      return response.data.map(sanitizeProduct);
    } catch (error) {
      console.error("Error searching products:", extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "SEARCH_PRODUCTS_ERROR"
      );
    }
  },

  // Get products by category
  getProductsByCategory: async (categoryId: string): Promise<Product[]> => {
    if (!categoryId || categoryId.trim() === "") {
      throw new ApiError("Category ID is required", 400, "INVALID_CATEGORY_ID");
    }
    
    try {
      const response = await apiClient.get<Product[]>(
        `/products/category/${encodeURIComponent(categoryId)}`
      );
      
      if (!isValidProductArray(response.data)) {
        console.warn("Invalid product data received from category API");
        return [];
      }
      
      return response.data.map(sanitizeProduct);
    } catch (error) {
      console.error(`Error fetching products for category ${categoryId}:`, extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "FETCH_CATEGORY_PRODUCTS_ERROR"
      );
    }
  },

  // Get products by price range
  getProductsByPriceRange: async (minPrice: number, maxPrice: number): Promise<Product[]> => {
    // Validate and sanitize price range
    let sanitizedMin = Math.max(0, minPrice || 0);
    let sanitizedMax = Math.max(0, maxPrice || Number.MAX_SAFE_INTEGER);
    
    // Swap if min > max
    if (sanitizedMin > sanitizedMax) {
      [sanitizedMin, sanitizedMax] = [sanitizedMax, sanitizedMin];
    }
    
    try {
      const response = await apiClient.get<Product[]>("/products/price-range", {
        params: { minPrice: sanitizedMin, maxPrice: sanitizedMax },
      });
      
      if (!isValidProductArray(response.data)) {
        console.warn("Invalid product data received from price range API");
        return [];
      }
      
      return response.data.map(sanitizeProduct);
    } catch (error) {
      console.error("Error fetching products by price range:", extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "FETCH_PRICE_RANGE_ERROR"
      );
    }
  },

  // Create a new product (admin)
  createProduct: async (product: Omit<Product, 'id' | 'createdAt' | 'updatedAt'>): Promise<Product> => {
    // Validate required fields
    if (!product.name?.trim()) {
      throw new ApiError("Product name is required", 400, "VALIDATION_ERROR");
    }
    if (product.price === undefined || product.price <= 0) {
      throw new ApiError("Product price must be greater than 0", 400, "VALIDATION_ERROR");
    }
    if (product.stock === undefined || product.stock < 0) {
      throw new ApiError("Product stock cannot be negative", 400, "VALIDATION_ERROR");
    }
    
    try {
      const sanitizedProduct = {
        ...product,
        name: product.name.trim(),
        description: product.description?.trim() || "",
        price: Math.max(0.01, product.price),
        stock: Math.max(0, product.stock),
      };
      
      const response = await apiClient.post<Product>("/products", sanitizedProduct);
      
      if (!isValidProduct(response.data)) {
        throw new ApiError("Invalid product data received", 500, "INVALID_DATA");
      }
      
      return sanitizeProduct(response.data);
    } catch (error) {
      if (error instanceof ApiError) throw error;
      console.error("Error creating product:", extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "CREATE_PRODUCT_ERROR"
      );
    }
  },

  // Update a product (admin)
  updateProduct: async (id: string, product: Partial<Product>): Promise<Product> => {
    if (!id || id.trim() === "") {
      throw new ApiError("Product ID is required", 400, "INVALID_ID");
    }
    
    try {
      const sanitizedProduct = {
        ...product,
        name: product.name?.trim(),
        description: product.description?.trim(),
        price: product.price !== undefined ? Math.max(0.01, product.price) : undefined,
        stock: product.stock !== undefined ? Math.max(0, product.stock) : undefined,
      };
      
      const response = await apiClient.put<Product>(
        `/products/${encodeURIComponent(id)}`,
        sanitizedProduct
      );
      
      if (!isValidProduct(response.data)) {
        throw new ApiError("Invalid product data received", 500, "INVALID_DATA");
      }
      
      return sanitizeProduct(response.data);
    } catch (error) {
      if (error instanceof ApiError) throw error;
      console.error(`Error updating product ${id}:`, extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "UPDATE_PRODUCT_ERROR"
      );
    }
  },

  // Delete a product (admin)
  deleteProduct: async (id: string): Promise<void> => {
    if (!id || id.trim() === "") {
      throw new ApiError("Product ID is required", 400, "INVALID_ID");
    }
    
    try {
      await apiClient.delete(`/products/${encodeURIComponent(id)}`);
    } catch (error) {
      console.error(`Error deleting product ${id}:`, extractErrorMessage(error));
      throw new ApiError(
        extractErrorMessage(error),
        error instanceof ApiError ? error.status : 500,
        "DELETE_PRODUCT_ERROR"
      );
    }
  },
};

File: ecommerce-core-service/src/main/java/com/ecommerce/core/service/ProductService.java
package com.ecommerce.core.service;

import com.ecommerce.core.dto.ProductDTO;
import com.ecommerce.core.exception.ResourceNotFoundException;
import com.ecommerce.core.exception.ValidationException;
import com.ecommerce.core.model.Product;
import com.ecommerce.core.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;

    public ProductDTO getProductById(Long id) {
        if (id == null || id <= 0) {
            throw new ValidationException("Product ID must be a positive number");
        }
        log.debug("Fetching product with id: {}", id);
        Product product = productRepository.findByIdAndActiveTrue(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        return convertToDTO(product);
    }

    public List<ProductDTO> getAllProducts() {
        log.debug("Fetching all products");
        List<Product> products = productRepository.findByActiveTrue();
        if (products == null || products.isEmpty()) {
            log.info("No active products found");
            return Collections.emptyList();
        }
        return products.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public List<ProductDTO> getProductsByCategory(Long categoryId) {
        if (categoryId == null || categoryId <= 0) {
            throw new ValidationException("Category ID must be a positive number");
        }
        log.debug("Fetching products for category: {}", categoryId);
        List<Product> products = productRepository.findByCategoryIdAndActiveTrue(categoryId);
        if (products == null || products.isEmpty()) {
            log.info("No products found for category: {}", categoryId);
            return Collections.emptyList();
        }
        return products.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public List<ProductDTO> searchProducts(String query) {
        log.debug("Searching products with query: {}", query);
        // Handle empty or null query
        if (!StringUtils.hasText(query)) {
            log.info("Empty search query, returning all products");
            return getAllProducts();
        }
        // Sanitize query - remove potential SQL injection characters
        String sanitizedQuery = query.trim()
                .replaceAll("[<>\"'%;()&+]", "")
                .substring(0, Math.min(query.length(), 200));

        List<Product> products = productRepository.searchProducts(sanitizedQuery);
        if (products == null || products.isEmpty()) {
            log.info("No products found for query: {}", sanitizedQuery);
            return Collections.emptyList();
        }
        return products.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    public List<ProductDTO> findByPriceRange(Double minPrice, Double maxPrice) {
        // Validate and sanitize price range
        if (minPrice == null) {
            minPrice = 0.0;
        }
        if (maxPrice == null) {
            maxPrice = Double.MAX_VALUE;
        }
        if (minPrice < 0) {
            throw new ValidationException("Minimum price cannot be negative");
        }
        if (maxPrice < 0) {
            throw new ValidationException("Maximum price cannot be negative");
        }
        if (minPrice > maxPrice) {
            // Swap values if min > max
            Double temp = minPrice;
            minPrice = maxPrice;
            maxPrice = temp;
            log.info("Swapped price range values: {} - {}", minPrice, maxPrice);
        }

        log.debug("Fetching products in price range: {} - {}", minPrice, maxPrice);
        List<Product> products = productRepository.findByPriceRange(minPrice, maxPrice);
        if (products == null || products.isEmpty()) {
            log.info("No products found in price range: {} - {}", minPrice, maxPrice);
            return Collections.emptyList();
        }
        return products.stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    @Transactional
    public ProductDTO createProduct(ProductDTO productDTO) {
        validateProductDTO(productDTO);

        log.debug("Creating new product: {}", productDTO.getName());
        Product product = Product.builder()
                .name(productDTO.getName().trim())
                .description(productDTO.getDescription() != null ? productDTO.getDescription().trim() : null)
                .price(productDTO.getPrice())
                .categoryId(productDTO.getCategoryId())
                .image(productDTO.getImage() != null ? productDTO.getImage().trim() : null)
                .stock(productDTO.getStock() != null ? productDTO.getStock() : 0)
                .rating(productDTO.getRating() != null ? Math.min(5.0, Math.max(0.0, productDTO.getRating())) : 0.0)
                .active(true)
                .build();
        Product savedProduct = productRepository.save(product);
        log.info("Created product with id: {}", savedProduct.getId());
        return convertToDTO(savedProduct);
    }

    @Transactional
    public ProductDTO updateProduct(Long id, ProductDTO productDTO) {
        if (id == null || id <= 0) {
            throw new ValidationException("Product ID must be a positive number");
        }
        validateProductDTO(productDTO);

        log.debug("Updating product with id: {}", id);
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));

        product.setName(productDTO.getName().trim());
        product.setDescription(
                productDTO.getDescription() != null ? productDTO.getDescription().trim() : product.getDescription());
        product.setPrice(productDTO.getPrice());
        product.setCategoryId(productDTO.getCategoryId());
        product.setImage(productDTO.getImage() != null ? productDTO.getImage().trim() : product.getImage());
        product.setStock(productDTO.getStock() != null ? productDTO.getStock() : product.getStock());
        if (productDTO.getRating() != null) {
            product.setRating(Math.min(5.0, Math.max(0.0, productDTO.getRating())));
        }

        Product updatedProduct = productRepository.save(product);
        log.info("Updated product with id: {}", updatedProduct.getId());
        return convertToDTO(updatedProduct);
    }

    @Transactional
    public void deleteProduct(Long id) {
        if (id == null || id <= 0) {
            throw new ValidationException("Product ID must be a positive number");
        }
        log.debug("Deleting product with id: {}", id);
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + id));
        product.setActive(false);
        productRepository.save(product);
        log.info("Soft deleted product with id: {}", id);
    }

    private void validateProductDTO(ProductDTO dto) {
        if (dto == null) {
            throw new ValidationException("Product data is required");
        }
        if (!StringUtils.hasText(dto.getName())) {
            throw new ValidationException("Product name is required");
        }
        if (dto.getName().length() > 255) {
            throw new ValidationException("Product name must be less than 255 characters");
        }
        if (dto.getPrice() == null || dto.getPrice() <= 0) {
            throw new ValidationException("Product price must be greater than 0");
        }
        if (dto.getPrice() > 999999.99) {
            throw new ValidationException("Product price cannot exceed 999,999.99");
        }
        if (dto.getStock() != null && dto.getStock() < 0) {
            throw new ValidationException("Stock cannot be negative");
        }
        if (dto.getDescription() != null && dto.getDescription().length() > 5000) {
            throw new ValidationException("Description must be less than 5000 characters");
        }
    }

    private ProductDTO convertToDTO(Product product) {
        if (product == null) {
            return null;
        }
        return ProductDTO.builder()
                .id(product.getId())
                .name(product.getName())
                .description(product.getDescription())
                .price(product.getPrice())
                .categoryId(product.getCategoryId())
                .image(product.getImage())
                .stock(product.getStock())
                .rating(product.getRating())
                .active(product.getActive())
                .createdAt(product.getCreatedAt())
                .updatedAt(product.getUpdatedAt())
                .build();
    }
}
